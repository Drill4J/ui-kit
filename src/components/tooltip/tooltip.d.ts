/// <reference types="react" />
export interface Props {
  className?: string;
  message: React.ReactNode;
  children: React.ReactNode | React.ReactNode[];
  position?: 'top-center' | 'top-right' | 'top-left' | 'left' | 'right';
}
export declare const Tooltip: {
  ({
    className, message, children, position,
  }: Props): JSX.Element;
  displayName: string;
  __docgenInfo: {
    description: string;
    displayName: string;
    props: {
      className: {
        defaultValue: null;
        description: string;
        name: string;
        required: boolean;
        type: {
          name: string;
        };
      };
      message: {
        defaultValue: null;
        description: string;
        name: string;
        required: boolean;
        type: {
          name: string;
        };
      };
      position: {
        defaultValue: {
          value: string;
        };
        description: string;
        name: string;
        required: boolean;
        type: {
          name: string;
        };
      };
    };
  };
};
