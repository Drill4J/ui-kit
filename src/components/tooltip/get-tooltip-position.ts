interface Anchors {
  childrenTopPosition: number;
  childrenLeftPosition: number;
  childrenWidth: number;
  childrenHeight: number;
  messageHeight: number;
  messageWidth: number;
  offset: number;
}

export const getTooltipPosition = (tooltipPositionType: string, anchors: Anchors) => {
  const isInvalidAnchors = Object.values(anchors).some(anchor => Number.isNaN(anchor) || anchor === Infinity);
  const {
    childrenTopPosition, childrenLeftPosition, childrenWidth, childrenHeight, messageHeight, messageWidth, offset,
  } = anchors;

  const topPosition = `${childrenTopPosition - messageHeight - offset}px`;
  switch (isInvalidAnchors ? 'default' : tooltipPositionType) {
    case 'top-center':
      return {
        top: topPosition,
        left: `${childrenLeftPosition - (messageWidth - childrenWidth) / 2}px`,
      };
    case 'top-right':
      return {
        top: topPosition,
        left: `${childrenLeftPosition + childrenWidth / 2}px`,
      };
    case 'top-left':
      return {
        top: topPosition,
        left: `${childrenLeftPosition - messageWidth + childrenWidth / 2}px`,
      };
    case 'right':
      return {
        top: `${childrenTopPosition + (childrenHeight - messageHeight) / 2}px`,
        left: `${childrenLeftPosition + childrenWidth + offset}px`,
      };
    case 'left':
      return {
        top: `${childrenTopPosition + (childrenHeight - messageHeight) / 2}px`,
        left: `${childrenLeftPosition - messageWidth - offset}px`,
      };
    default:
      return {
        top: `${0}px`,
        left: `${0}px`,
      };
  }
};
