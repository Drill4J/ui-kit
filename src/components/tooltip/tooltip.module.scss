@import '../../style-guide/index';

.tooltip {
  display: inline-block;

  &__message {
    position: absolute;
    white-space: pre;
    text-decoration: none;
    color: map-get($monochrome, white);
    font-size: 12px;
    background: map-get($monochrome, black);
    border-radius: 4px;
    z-index: 100;

    & > div {
      padding: 8px 16px;
    }

    &::after {
      position: absolute;
      border: 8px solid transparent;
      content: '';
    }

    &--type {
      &--top-center {
        &::after {
          bottom: -8px;
          right: calc(50% - 8px);
          border-top-color: map-get($monochrome, black);
          border-bottom: none;
        }
      }

      &--top-right {
        &::after {
          bottom: -8px;
          right: calc(100% - 8px);
          border-left-color: map-get($monochrome, black);
          border-right: none;
        }
      }

      &--top-left {
        &::after {
          bottom: -8px;
          left: calc(100% - 8px);
          border-right-color: map-get($monochrome, black);
          border-left: none;
        }
      }

      &--right {
        &::after {
          bottom: 8px;
          right: 100%;
          border-right-color: map-get($monochrome, black);
          border-left: none;
        }
      }

      &--left {
        &::after {
          bottom: 8px;
          left: 100%;
          border-left-color: map-get($monochrome, black);
          border-right: none;
        }
      }
    }
  }
}
