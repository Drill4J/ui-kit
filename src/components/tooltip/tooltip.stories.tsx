import { storiesOf } from '@storybook/react';
import 'twin.macro';

import { Tooltip } from './tooltip';

storiesOf('Tooltip', module).add('Tooltip', () => (
  <>
    <div tw="my-10 mx-20">
      <Tooltip message="Lorem ipsum dolor sit amet, consectetur adipisicing elit. Nam vero">
        Tooltip
        <div>top-right</div>
      </Tooltip>
    </div>
    <div tw="my-16 mx-30">
      <Tooltip
        message={(
          <>
            <div>
              <div>
                <div>Lorem ipsum dolor sit amet</div>
                <div>Lorem ipsum dolor sit amet</div>
              </div>
              <div>Consequatur rerum amet fuga expedita</div>
            </div>
            <div>
              <div>
                <div>Lorem ipsum dolor sit amet</div>
                <div>Lorem ipsum dolor sit amet</div>
              </div>
              <div>Consequatur rerum amet fuga expedita</div>
            </div>
          </>
        )}
      >
        Tooltip
        <div>top-center</div>
      </Tooltip>
    </div>
    <div tw="my-20 mx-100">
      <Tooltip
        position="right"
        message="Consequatur rerum amet fuga expedita sunt et tempora saepe? Iusto nihil explicabo..."
      >
        Tooltip
        <div>right</div>
      </Tooltip>
    </div>
    <div tw="my-20 mx-100">
      <Tooltip
        position="left"
        message="Consequatur rerum amet fuga expedita sunt et tempora saepe? Iusto nihil explicabo..."
      >
        Tooltip
        <div>left</div>
      </Tooltip>
    </div>
    <div tw="my-20 mx-100">
      <Tooltip
        position="top-left"
        message="Consequatur rerum amet fuga expedita sunt et tempora saepe? Iusto nihil explicabo..."
      >
        Tooltip
        <div>top-left</div>
      </Tooltip>
    </div>
  </>
));
