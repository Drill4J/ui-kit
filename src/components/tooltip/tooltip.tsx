import { useLayoutEffect, useRef, useState } from 'react';
import styled, { css } from 'styled-components';

import { useHover } from '../../hooks';
import { Portal } from '../portal';
import { getTooltipPosition } from './get-tooltip-position';
import { COLORS } from '../../theme';

export interface Props {
  className?: string;
  message: React.ReactNode;
  children: React.ReactNode | React.ReactNode[];
  position?: 'top-center' | 'top-right' | 'top-left' | 'left' | 'right';
}

export const Tooltip = ({
  className, message, children, position = 'top-center',
}: Props) => {
  const offset = 12;
  const { ref, isVisible } = useHover();

  const [tooltipPositionType, setTooltipPositionType] = useState(position);
  const [{ height: messageHeight, width: messageWidth }, setMessageCoords] = useState({
    height: 0,
    width: 0,
  } as DOMRect);
  const messageRef = useRef<HTMLDivElement>(null);
  const childrenRef = useRef<HTMLDivElement>(null);

  useLayoutEffect(() => {
    const messageCoords = messageRef?.current?.getBoundingClientRect();
    isVisible && messageCoords && setMessageCoords(messageCoords);

    const observer = new IntersectionObserver(
      ([entry]) =>
        !entry.isIntersecting &&
          setTooltipPositionType(entry.intersectionRect.left ? 'top-left' : 'top-right'),
      {
        root: null,
        threshold: 1.0,
      },
    );
    messageRef.current && observer.observe(messageRef.current);

    return () => {
      observer.disconnect();
    };
  }, [isVisible]);

  const {
    top: childrenTopPosition = 0,
    left: childrenLeftPosition = 0,
    width: childrenWidth = 0,
    height: childrenHeight = 0,
  } = childrenRef?.current?.getBoundingClientRect() || {};

  const anchors = {
    childrenTopPosition,
    childrenLeftPosition,
    childrenWidth,
    childrenHeight,
    messageHeight,
    messageWidth,
    offset,
  };

  return (
    <Wrapper className={className}>
      <div ref={ref}>
        <div ref={childrenRef}>{children}</div>
      </div>
      {isVisible && message && (
        <Portal rootElementId="tooltip">
          <Message
            style={getTooltipPosition(tooltipPositionType, anchors)}
            type={tooltipPositionType}
          >
            <div ref={messageRef}>{message}</div>
          </Message>
        </Portal>
      )}
    </Wrapper>
  );
};

const Wrapper = styled.div`
  display: inline-block;
`;

const Message = styled.div<{type?: 'top-center' | 'top-right' | 'top-left' | 'left' | 'right'}>`
  position: absolute;
  white-space: pre;
  text-decoration: none;
  color: ${COLORS.MONOCHROME.WHITE};
  font-size: 12px;
  background: ${COLORS.MONOCHROME.BLACK};
  border-radius: 4px;
  z-index: 100;

  & > div {
    padding: 8px 16px;
  }

  &::after {
    width: 8px;
    height: 8px;
    position: absolute;
    border: 8px solid transparent;
    content: '';
    ${({ type }) => [
    type === 'top-left' && css`
      bottom: -8px;
      left: calc(100% - 8px);
      border-right-color: ${COLORS.MONOCHROME.BLACK};
      border-left: none;
    `,
    type === 'top-center' && css`
      bottom: -8px;
      right: calc(50% - 8px);
      border-top-color: ${COLORS.MONOCHROME.BLACK};
      border-bottom: none;
    `,
    type === 'top-right' && css`
      bottom: -8px;
      right: calc(100% - 8px);
      border-left-color: ${COLORS.MONOCHROME.BLACK};
      border-right: none;
    `,
    type === 'left' && css`
      bottom: 8px;
      left: 100%;
      border-left-color: ${COLORS.MONOCHROME.BLACK};
      border-right: none;
    `,
    type === 'right' && css`
      bottom: 8px;
      right: 100%;
      border-right-color: ${COLORS.MONOCHROME.BLACK};
      border-left: none;
    `,
  ]}
  }
`;
