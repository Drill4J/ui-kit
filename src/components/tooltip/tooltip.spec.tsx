import { shallow, ShallowWrapper } from 'enzyme';

import { Tooltip } from './tooltip';

describe('Tooltip', () => {
  let wrapper: ShallowWrapper;
  beforeAll(() => {
    wrapper = shallow(
      <Tooltip message={<span>Tooltip</span>}>
        <div>Tooltip</div>
      </Tooltip>,
      {
        disableLifecycleMethods: true,
      },
    );
  });

  it('should match snapshot', () => {
    expect(wrapper.getElement()).toMatchSnapshot();
  });
});
