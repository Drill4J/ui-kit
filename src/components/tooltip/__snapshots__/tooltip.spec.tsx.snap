// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip should match snapshot 1`] = `
<styled.div>
  <div>
    <div>
      <div>
        Tooltip
      </div>
    </div>
  </div>
</styled.div>
`;
